fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,492 @@
+
+/// Secp256r1 public key.
+#[readonly::make]
+#[derive(Debug, Clone)]
+pub struct Secp256r1PublicKey {
+    pub pubkey: ExternalPublicKey,
+    pub bytes: OnceCell<[u8; PUBLIC_KEY_SIZE]>,
+}
+
+/// Binary representation of an instance of [Secp256r1PublicKey].
+pub type Secp256r1PublicKeyBytes = PublicKeyBytes<Secp256r1PublicKey, { PUBLIC_KEY_SIZE }>;
+
+/// Secp256r1 private key.
+#[readonly::make]
+#[derive(SilentDebug, SilentDisplay)]
+pub struct Secp256r1PrivateKey {
+    pub privkey: ExternalSecretKey,
+    pub bytes: OnceCell<[u8; PRIVATE_KEY_SIZE]>,
+}
+
+/// Secp256r1 signature.
+#[readonly::make]
+#[derive(Debug, Clone)]
+pub struct Secp256r1Signature {
+    pub sig: ExternalSignature,
+    pub bytes: OnceCell<[u8; SIGNATURE_SIZE]>,
+}
+
+impl std::hash::Hash for Secp256r1PublicKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl PartialOrd for Secp256r1PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.as_ref().partial_cmp(other.as_ref())
+    }
+}
+
+impl Ord for Secp256r1PublicKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_ref().cmp(other.as_ref())
+    }
+}
+
+impl PartialEq for Secp256r1PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.pubkey == other.pubkey
+    }
+}
+
+impl Eq for Secp256r1PublicKey {}
+
+impl VerifyingKey for Secp256r1PublicKey {
+    type PrivKey = Secp256r1PrivateKey;
+    type Sig = Secp256r1Signature;
+    const LENGTH: usize = PUBLIC_KEY_SIZE;
+}
+
+impl Verifier<Secp256r1Signature> for Secp256r1PublicKey {
+    fn verify(&self, msg: &[u8], signature: &Secp256r1Signature) -> Result<(), signature::Error> {
+        // TODO: Until we have a recovery id, we recover both candidates for public keys
+        let pks = signature
+            .recover(msg)
+            .map_err(|_| signature::Error::new())?;
+
+        if !pks.contains(self) {
+            return Err(signature::Error::new());
+        }
+
+        self.pubkey
+            .verify(msg, &signature.sig)
+            .map_err(|_| signature::Error::new())
+    }
+}
+
+impl Secp256r1PublicKey {}
+
+impl AsRef<[u8]> for Secp256r1PublicKey {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| Ok(self.pubkey.as_ref().to_bytes().into()))
+            .expect("OnceCell invariant violated")
+    }
+}
+
+impl ToFromBytes for Secp256r1PublicKey {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FastCryptoError> {
+        match ExternalPublicKey::try_from(bytes) {
+            Ok(pubkey) => Ok(Secp256r1PublicKey {
+                pubkey,
+                bytes: OnceCell::new(),
+            }),
+            Err(_) => Err(FastCryptoError::InvalidInput),
+        }
+    }
+}
+
+impl Default for Secp256r1PublicKey {
+    fn default() -> Self {
+        Secp256r1PublicKey::from_bytes(&[0u8; PUBLIC_KEY_SIZE]).unwrap()
+    }
+}
+
```

Comment:
This will 100% panic if called. Please fix, test. I wonder if we've made the same mistake on secp256k1?

Reply:
We did. I'll create an issue.